### ISSUE TITLE
Differentiable shuffle function

### ISSUE THREAD

**Issue by joshnunley** [FluxML/Flux.jl#1766]
I may be misunderstanding something, but I would like to be able to differentiate through Julia's shuffle function on vectors. The function acting on type Vector{Float64} is essentially just multiplication by a permutation matrix, but generating, then multiplying by a permutation matrix is significantly slower than shuffling the vector. Can I somehow define a custom adjoint, something like 

`@adjoint shuffle(a::Vector{Float64}) = shuffle(a), d -> d*shuffle(a)`

?

**Comment by ToucheSir** [comment]
I'm not sure I completely understand the question. Taking gradients with Flux doesn't update parameters outside of any mutations you yourself make on the forward pass (e.g. BatchNorm moving average updates). We include gradient-based optimizers under `Flux.Optimise` because that's what most folks use for their NNs, but you can manipulate weights however you want outside of the gradient context.

**Comment by joshnunley** [comment]
Here is a simple example of what I want to do:

`
mutable struct PermutationModel
    model::Any
    parameters::Vector{Float64}
end

function (m::PermutationModel)(x)
    p, restruct = Flux.destructure(m.model)
    
    shuffle!(m.parameters)
    
    reparam_model = restruct(m.parameters)
    return reparam_model(x)
end`

I just want Flux to update the parameters with the gradient of the model at the given parameters, regardless of whether the way in which the parameters were generated was differentiable.

**Comment by ToucheSir** [comment]
Well yes, AD is not magic. Zygote in particular operates on a non-optimized IR and thus can not simply eliminate any variables that might not contribute to the final result (for all it knows, those two lines might be modifying `p` behind the scenes).

It would help to get some context on what you're trying to do, as I assume the example above is not it. Generally speaking though, it's best to hoist anything which doesn't require differentiation outside of the `gradient` callback. e.g.
```julia
p, restruct = Flux.destructure(model)

# for batch in data ...
shuffle!(p)
gs = gradient(p) do ps
  reparam_model = restruct(ps)
  loss(reparam_model(...), ...)
end
```

**Comment by joshnunley** [comment]
Thank you for your help so far! I am a fairly new user of Flux. My minimal example is actually fairly similar to my actual use case. I would like to be able to swap some parameters, calculate the gradient with respect to the swapped parameters, then apply the gradient to the unswapped (original) parameters. Ideally I could do this in a nice way, so that I could sum the loss over many different random parameter swaps, calculate the gradient of that summed loss with respect to the original parameters, and apply that gradient to the original unswapped parameters.

**Comment by ToucheSir** [comment]
Correct me if I'm wrong, but it seems you could do this in a kind of gradient accumulation style? e.g.
```julia
p, restruct = Flux.destructure(model)
accum_gs = fill!(similar(p), 0) # gradient buffer

# for batch in data ...
for _ in 1:nswaps
  # substitute in-place versions and other optimizations as desired
  swapped_p = swap(p, ...)
  gs = gradient(swapped_p) do ps
    reparam_model = restruct(ps)
    loss(reparam_model(x), y)
  end
  accum_gs .+= unswap(gs, ...)
end

# update p with accum_gs
```

**Comment by joshnunley** [comment]
Thank you again for taking the time to write this out. We are getting closer but I think this won't quite work, though it's not entirely obvious. Accumulating the gradient with mini-batches works because here it turns out that the gradient of the summed losses of each mini-batch is the same as the summed gradient over the losses of the mini-batches.

Unfortunately that may not be the case with parameter swaps. It helps to imagine there is a single, definite scalar field defining the loss at every point in parameter space (this is the thing we calculate stochastically with batches). In my case, I would like to sum up the loss at different points in parameter space (the parameter swaps), then calculate the gradient of that summed loss and apply it to my model with unswapped parameters. If I take the gradient at each parameter-swapped location in parameter space, then add them, there is no guarantee that this is related to the gradient of the summed loss.

**Comment by ToucheSir** [comment]
Don't feel obligated to respond quickly! Certainly neither I nor most of the other contributors would complain about having a quieter GH inbox :smile:.

The reason I suggested the approach above is that (assuming it works), it would be the path of least resistance. It is possible to have the swapping take part in a gradient context, but that would be a far more involved process (you are almost guaranteed to have to write custom rules to appease the AD) and likely far less efficient as well.
